include/deal.II/matrix_free/cuda_tensor_product_kernels.cuh

```diff
@@ -0,0 +1,395 @@
+
+      /**
+       * Helper function for values() and gradients().
+       */
+      template <int direction, bool dof_to_quad, bool add, bool in_place>
+      __device__ void apply(Number shape_data[],
+                            const Number *in,
+                            Number       *out) const;
+
+      /**
+       * Evaluate the finite element function at the quadrature points.
+       */
+      __device__ void value_at_quad_pts(Number *u);
+
+      /**
+       * Helper function for integrate(). Integrate the finite element function.
+       */
+      __device__ void integrate_value(Number *u);
+
+      /**
+       * Evaluate the gradients of the finite element function at the quadrature
+       * points.
+       */
+      __device__ void gradient_at_quad_pts(const Number *const u,
+                                           Number *grad_u[dim]);
+
+      /**
+       * Helper function for integrate(). Integrate the gradients of the finite
+       * element function.
+       */
+      template <bool add>
+      __device__ void integrate_gradient(Number *u,
+                                         Number *grad_u[dim]);
+    };
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    __device__ EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::EvaluatorTensorProduct()
+    {}
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <int direction, bool dof_to_quad, bool add, bool in_place>
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::values(const Number *in,
+                                              Number *out) const
+    {
+      apply<direction, dof_to_quad, add, in_place>(global_shape_values, in, out);
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <int direction, bool dof_to_quad, bool add, bool in_place>
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::gradients(const Number *in,
+                                                 Number *out) const
+    {
+      apply<direction, dof_to_quad, add, in_place>(global_shape_gradients, in, out);
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <int direction, bool dof_to_quad, bool add, bool in_place>
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::apply(Number shape_data[],
+                                             const Number *in,
+                                             Number       *out) const
+    {
+      const unsigned int i = (dim == 1) ? 0 : threadIdx.x%n_q_points_1d;
+      const unsigned int j = (dim == 3) ? threadIdx.y : 0;
+      const unsigned int q =
+        (dim == 1) ? (threadIdx.x%n_q_points_1d) :
```

Comment:
You can't assume that this code will perform exactly like what @kalj had because I did some modifications to it to make it "fit" better with the rest of the library. However, on this part this is exactly what @kalj did https://github.com/kalj/dealii-cuda/blob/710a46a6c2f98c966ad1842f18e29bcfbabbe704/matrix_free_gpu/tensor_ops.cuh#L26